### ISSUE TITLE
Feature request: FlashAttention

### ISSUE THREAD

**Issue by OhadRubin** [google/flax#2858]
Hey,
There is an implementation for [this](https://arxiv.org/abs/2205.14135) memory efficient version of attention [here](https://github.com/lucidrains/flash-attention-jax).
And I was wondering if it is possible to somehow integrate it into Flax.

**Comment by Lime-Cakes** [comment]
Instead of flash attention, maybe consider memory efficient attention? https://arxiv.org/abs/2112.05682

 I tested that flash attention and noticed it's both slower and seem to take more memory. Tests were done on TPU, so it's possible the higher memory usage is due to flash attention slicing causing more padding. 

Memory efficient attention have shown a decrease in memory. The paper actually comes with jax implementation. Google research's implementation can be found https://github.com/google-research/google-research/tree/master/memory_efficient_attention

**Comment by aniquetahir** [comment]
It's supposed to be both faster and less memory intensive. lucidrains version is implemented in jax instead of at a low level. So it is slower than the normal version.

**Comment by agemagician** [comment]
+1
Any update on this requst ?

**Comment by IvyZX** [comment]
`jax.experimental` has an [implementation](https://github.com/google/jax/blob/main/jax/experimental/pallas/ops/tpu/flash_attention.py) of FlashAttention, written by Pallas kernels and therefore usable in both GPU and TPU.

We can probably upstream this to Flax attention if `jax.experimental` doesn't scare us (it means the API could still change).

**Comment by aniquetahir** [comment]
> `jax.experimental` has an [implementation](https://github.com/google/jax/blob/main/jax/experimental/pallas/ops/tpu/flash_attention.py) of FlashAttention, written by Pallas kernels and therefore usable in both GPU and TPU.
> 
> We can probably upstream this to Flax attention if `jax.experimental` doesn't scare us (it means the API could still change).

Does that support both backdrop and forward prop?

**Comment by agemagician** [comment]
> > `jax.experimental` has an [implementation](https://github.com/google/jax/blob/main/jax/experimental/pallas/ops/tpu/flash_attention.py) of FlashAttention, written by Pallas kernels and therefore usable in both GPU and TPU.
> > We can probably upstream this to Flax attention if `jax.experimental` doesn't scare us (it means the API could still change).
> 
> Does that support both backprop and forward prop?

Yes, it seems so.

**Comment by agemagician** [comment]
> `jax.experimental` has an [implementation](https://github.com/google/jax/blob/main/jax/experimental/pallas/ops/tpu/flash_attention.py) of FlashAttention, written by Pallas kernels and therefore usable in both GPU and TPU.
> 
> We can probably upstream this to Flax attention if `jax.experimental` doesn't scare us (it means the API could still change).

Yes, but it seems missing some parameters like the dropout, mask and bias, if I am not mistaken.

**Comment by agemagician** [comment]
One more point Jax has two implementations:
1. Fused attention: https://github.com/google/jax/blob/main/jax/experimental/pallas/ops/attention.py
2. Flash attention: https://github.com/google/jax/blob/main/jax/experimental/pallas/ops/tpu/flash_attention.py

**Comment by Lime-Cakes** [comment]
Update: my comment was made before Pallas was an option, where low level access is simply not possible. 

The official jax flash attention implementation is built off on Pallas and thus not subjected to the same issue from 5+months ago. Flash attention should be variable right now, and likely faster than the older mem eff attention I suggested.

**Comment by IvyZX** [comment]
Looks like the two attention kernels are for different platforms - one for [TPU](https://github.com/google/jax/blob/main/jax/experimental/pallas/ops/tpu/flash_attention.py) and another for [GPU](https://github.com/google/jax/blob/main/jax/experimental/pallas/ops/attention.py). The implementations are slightly different for performance reasons, and they are also of different APIs.

It would be cool if we can unify the two and provide a stable attention API, either at Flax level or at JAX level. I will look more into that and collaborate with JAX folks.

For now, feel free to just use one of those kernels based on your platform needs.

**Comment by agemagician** [comment]
Thanks a lot @IvyZX  for integerating flash attention.

I am just afraid that it is still missing some parameters like the dropout, mask and bias, if I am not mistaken.

**Comment by agemagician** [comment]
@IvyZX  could you please share the link for the current PR ?

**Comment by IvyZX** [comment]
Dropout is not currently available in Pallas kernels, as it is yet to support PRNG keys. Causal masking can be turned on with `causal=True`, and the TPU version has attention bias (the arg `ab` - unclear naming IMO).

I wonder what the real-life use cases for attention bias are, as many LLMs nowadays doesn't use bias any more. Could anyone share why they would like to have it?

**Comment by agemagician** [comment]
If I am not mistaken, many of the state of the art Embedding method requires the attention bias like ALibi and relative attention encoding for T5 models.

**Comment by OhadRubin** [comment]
It seems that this triton implementation supports attention bias, so there is nothing that prevents the algorithm from supporting it.
https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/flash_attn_triton.py
Additionally, this implementation of block-parallel attention (faster than FA-1) also supports attention bias:
https://github.com/young-geng/EasyLM/blob/main/EasyLM/bpt.py

**Comment by maxencefaldor** [comment]
Hi,

Any news on this? It would be great to have flash attention in Flax.

**Comment by itsnamgyu** [comment]
Is the [jax.nn.dot_product_attention](https://docs.jax.dev/en/latest/_autosummary/jax.nn.dot_product_attention.html#jax.nn.dot_product_attention) implementation "highly" optimized for TPU? It doesn't seem to refer to any custom pallas kernels.

Or, should TPU users use the pallas version directly?

@IvyZX

**Comment by wchen99998** [comment]
> jax.nn.dot_product_attention

+1, I have the same question

**Comment by xiaobo-lab** [comment]
> Update as of 2025/05/27:
> 
> If you use [`flax.nnx.dot_product_attention`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/nn/attention.html#flax.nnx.dot_product_attention) or [`flax.nnx.MultiHeadAttention`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/nn/attention.html#flax.nnx.MultiHeadAttention) without dropout and with `sow_weights=False`, and you are on GPU, you should already be using Flash Attention kernel.
> 
> This is because Flax by default uses [`jax.nn.dot_product_attention`](https://docs.jax.dev/en/latest/_autosummary/jax.nn.dot_product_attention.html#jax.nn.dot_product_attention) which will automatically switch to the `cudnn` implementation.
> 
> Let us know if this failed to happen or if it doesn't cover your use case. Sorry about the wait!

no, is not working for me, when i run code below, i got OOM error

```python3
import jax
import time
import optax
import jax
import jax.numpy as jnp
import flax.linen as nn
from tqdm import tqdm
from flax import nnx
from flax.training import train_state


class MyModel(nn.Module):
    @nn.compact
    def __call__(self, external: jax.Array, internal: jax.Array, rngs: jax.Array):
        attention = nnx.MultiHeadAttention(4, 1024, rngs=nnx.Rngs(0), decode=False)(external, internal, internal)
        # attention = MultiHeadAttention(4, 5, 5, False, 0.1, jnp.float16)(external, internal, internal, deterministic)
        x = jnp.mean(attention, 1)
        x = nn.Dense(1)(x)
        return x


def focal_loss(logits: jax.Array, labels: jax.Array, alpha: float, gamma: float = 2.):
    logits = nn.sigmoid(logits)
    at = alpha * labels + (1 - alpha) * (1 - labels)
    pt = logits * labels + (1 - logits) * (1 - labels)
    fl = - at * jnp.pow(1 - pt, gamma) * jnp.log(pt + 1e-8)
    return fl


@jax.jit
def train_step(state: train_state.TrainState, external: jax.Array, internal: jax.Array, labels: jax.Array, rngs: jax.Array):
    @jax.value_and_grad
    def forward(params):
        logits = state.apply_fn({'params': params}, external, internal, rngs)
        return jnp.mean(focal_loss(logits, labels, 0.8))

    loss, grads = forward(state.params)
    return loss, grads


def main():
    random_key = jax.random.PRNGKey(1)
    model = MyModel()

    state = train_state.TrainState.create(apply_fn=model.apply, params=model.init(random_key, jnp.ones((3, 12800, 1024), jnp.float16), jnp.ones((3, 12800, 1024), jnp.float16), True)['params'], tx=optax.adamw(1e-7))

    loss_list = []
    start = time.time()
    for epoch in tqdm(range(500)):
        random_key, sub_key = jax.random.split(random_key)
        external = jax.random.uniform(random_key, (3, 12800, 1024), jnp.float16)
        internal = jax.random.uniform(sub_key, (3, 12800, 1024), jnp.float16)
        labels = jnp.ones((3, 1))
        loss, grads = train_step(state, external, internal, labels, sub_key)
        loss_list.append(loss)
        state = state.apply_gradients(grads=grads)
    print(f'time={time.time() - start}')


if __name__ == '__main__':
    main()

```

**Comment by MiladInk** [comment]
> Update as of 2025/05/27:
> 
> If you use [`flax.nnx.dot_product_attention`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/nn/attention.html#flax.nnx.dot_product_attention) or [`flax.nnx.MultiHeadAttention`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/nn/attention.html#flax.nnx.MultiHeadAttention) without dropout and with `sow_weights=False`, and you are on GPU, you should already be using Flash Attention kernel.
> 
> This is because Flax by default uses [`jax.nn.dot_product_attention`](https://docs.jax.dev/en/latest/_autosummary/jax.nn.dot_product_attention.html#jax.nn.dot_product_attention) which will automatically switch to the `cudnn` implementation.
> 
> Let us know if this failed to happen or if it doesn't cover your use case. Sorry about the wait!

Am I using the flash attention if I am using `nn.MultiHeadDotProductAttention` which is from flax.linen and not the nnx? I think not as it would use the source mentioned[ here  ](https://flax.readthedocs.io/en/v0.6.10/_modules/flax/linen/attention.html#MultiHeadDotProductAttention) that does not use the  `jax.nn.dot_product_attention` but I am wondering if there is a simple way to use flash-attention in linen and not nnx.

**Comment by IvyZX** [comment]
> Is the [jax.nn.dot_product_attention](https://docs.jax.dev/en/latest/_autosummary/jax.nn.dot_product_attention.html#jax.nn.dot_product_attention) implementation "highly" optimized for TPU? It doesn't seem to refer to any custom pallas kernels.
> 
> Or, should TPU users use the pallas version directly?

`jax.nn.dot_product_attention` only has XLA version for TPU. For faster performance you might want to use the Pallas version directly, which has slightly different API (e.g., `mask` vs `segment_id` & `causal` to pass in masking info). You might need to fork the NNX attention layer and configure things a bit.

**Comment by IvyZX** [comment]
> > Update as of 2025/05/27:
> > If you use [`flax.nnx.dot_product_attention`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/nn/attention.html#flax.nnx.dot_product_attention) or [`flax.nnx.MultiHeadAttention`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/nn/attention.html#flax.nnx.MultiHeadAttention) without dropout and with `sow_weights=False`, and you are on GPU, you should already be using Flash Attention kernel.
> > This is because Flax by default uses [`jax.nn.dot_product_attention`](https://docs.jax.dev/en/latest/_autosummary/jax.nn.dot_product_attention.html#jax.nn.dot_product_attention) which will automatically switch to the `cudnn` implementation.
> > Let us know if this failed to happen or if it doesn't cover your use case. Sorry about the wait!
> 
> Am I using the flash attention if I am using `nn.MultiHeadDotProductAttention` which is from flax.linen and not the nnx? I think not as it would use the source mentioned[ here ](https://flax.readthedocs.io/en/v0.6.10/_modules/flax/linen/attention.html#MultiHeadDotProductAttention) that does not use the `jax.nn.dot_product_attention` but I am wondering if there is a simple way to use flash-attention in linen and not nnx.

You can override the specific attention function being used in the Linen layer. In this case, initialize it like: `model = nn.MultiHeadDotProductAttention(..., attention_fn=nnx.dot_product_attention)`
